float32 theta
float32 range
float32 psi_p